### ISSUE TITLE
Create low-allocation immutable Roaring bitmaps

### ISSUE THREAD

**Issue by lemire** [RoaringBitmap/roaring#264]
Some users have a many bitmaps on disk. Loading them up temporarily for operations is expensive due in part to the new to allocate new instances.

Related PR https://github.com/RoaringBitmap/roaring/pull/263

Also related to issue https://github.com/RoaringBitmap/roaring/issues/193

cc @guymolinari @jacksonrnewhouse

**Comment by joenall** [comment]
That is our use case. We have many bitmaps in many boltdbs on disk. Bitmaps have about 100M bits. The whole collection has a up to ~6B objects that are indexed, but none of the shards have more than 32 bit indices.

**Comment by lemire** [comment]
@joenall Thanks for chiming in. It is really important to document the use case to motivate a public solution.

**Comment by kevinconaway** [comment]
We have a similar use case where we create a lot of immutable bitmaps in memory and only read from / iterate them.

It would be great if there were improvements that could be made for immutable maps.

**Comment by lemire** [comment]
@kevinconaway Just checking... do you create them in this manner....?


```Go
newrb:= roaring.New()
newrb.ReadFrom(buf)
```

**Comment by kevinconaway** [comment]
No, for this use case we typically create a new `Bitmap`, add values to it and then never modify it again for its lifetime:

```go
rb := roaring.New()
rb.Add()
rb.Add()
...
rb.RunOptimize()
```

We notice these occupying a fair amount of heap space / heap objects so I was curious if there was a more optimized in-memory layout when we know that we won't modify the `Bitmap` again

Apologies if I interjected on an unrelated ticket here.

**Comment by lemire** [comment]
@kevinconaway 

> We notice these occupying a fair amount of heap space / heap objects so I was curious if there was a more optimized in-memory layout when we know that we won't modify the Bitmap again

If you sum up `GetSizeInBytes()` on all bitmaps, you should get something close to your actual memory usage (though your actual memory usage will be a bit higher due to various overheads). If that's not the case, then you should report the issue as performance bug.

**Comment by ajeetdsouza** [comment]
We were facing this problem at [Dgraph Labs](http://dgraph.io/) too, so we built a library called [sroar](https://github.com/dgraph-io/sroar) where the in-memory representation matches the on-disk representation (no need for marshalling / unmarshalling). Perhaps you will find this useful.

**Comment by lemire** [comment]
Regarding the last comment, I created an issue specifically for it at https://github.com/RoaringBitmap/roaring/issues/298

**Comment by kevinconaway** [comment]
>If you sum up GetSizeInBytes() on all bitmaps, you should get something close to your actual memory usage 

Apologies for the confusion, the main concern here is the number of objects on the heap (the various `*roaringArray`, `*arrayContainer` and `*runContainer16` values that occupy the heap)

**Comment by lemire** [comment]
> Apologies for the confusion, the main concern here is the number of objects on the heap (the various *roaringArray, *arrayContainer and *runContainer16 values that occupy the heap)

I realize that... hence my question. The `GetSizeInBytes()` function will not include this overhead. We can compare this rather theoretical measure with the actual memory usage. The difference is something that can be optimized away. We have done such analysis in the past and found that the overhead was small... meaning that the benefits were modest. It could be that the story is different in some importance cases, but optimizing without benchmarks is a fool's game. So we need good documentation on the issue first.

(To be clear, I am not arguing. I am just clarifying.)

**Comment by Oppen** [comment]
Right. I thought it was a format issue, that's why I suggested just recycling the old one instead of creating a new one. So it's just implementing a new load?

**Comment by Oppen** [comment]
There are performance benefits on using aligned data AFAIR, tho. Reads and writes of aligned data should hit the bus less often. Same way, they should invalidate the cache less often. I assumed that's why we store stuff aligned in a way it matches natural alignment AVX2 registers for bitsets.

**Comment by lemire** [comment]
> There are performance benefits on using aligned data AFAIR, tho. Reads and writes of aligned data should hit the bus less often. Same way, they should invalidate the cache less often. I assumed that's why we store stuff aligned in a way it matches natural alignment AVX2 registers for bitsets.

Yes. People often claim that alignment is important for performance. Back in 2012, I wrote a blog post on the topic...

Data alignment for speed: myth or reality?
https://lemire.me/blog/2012/05/31/data-alignment-for-speed-myth-or-reality/

If you are scanning an array, you might hit one extra cache line due to lack of alignment. So you can build relatively contrived examples where this will impact your overall performance... but my claim is that by the time it is a concern, you will have hit many more low-hanging fruit.


It comes down to having simpler code. But if I could wrap the byte arrays in C with memcpy calls, without going insane, then we could do away entirely with the frozen format.

In C++, I could have a wrapper that calls memcpy... but in C, I'd probably need macros, and then it would get ugly.

**Comment by Oppen** [comment]
But if you do `memcpy` you lose part of the benefits of using a frozen format in the first place, as a big one is the ability to simply `mmap` and view. The other benefit being no extra processing nor too many allocations, however, is still intact.
But true, the cache hit argument is a bit moot if you scan the whole array. What about loading many files with `mmap` and computing `and`? In that case you could simply not load most of the bitmap, while having the view "complete" to the program's knowledge. The number of pages you'd load in that case for bitmaps could be up to 50% higher I think for not being aligned. It's contrived but I think it's possible, and I think it's actually a likely case for the project I use it in.
I'll read the article, it sounds interesting.

**Comment by Oppen** [comment]
Off-topic: there's a case where it really is more important to align to cache, when atomics are involved. Atomic operations invalidate a cache line on all processors, so you want to avoid those invalidating more than one line _and_ invalidating non-related data (so you usually try to fit more related data in the same line so it's really invalid for others anyway).

**Comment by lemire** [comment]
> But if you do memcpy you lose part of the benefits of using a frozen format in the first place, as a big one is the ability to simply mmap and view.

You do not.

You memcpy the individual words which brings them into register from an mmap pointer. You do not have to memcpy the array. There is no difference between `x[i]` and `uint64_t x; memcpy(&x, ..,)`. It compiles to the same code, down to the autovectorization.


> The number of pages you'd load in that case for bitmaps could be up to 50% higher I think for not being aligned.

If you are loading individual words and you put each of them at a page boundary, every other page, then you could, theoretically, have to double the number of pages you must load. But I can also come up with pathological case where alignment triggers massive cache aliasing issues (where all addresses have the same least significant bits) that destroys your cache. Pathological cases should not guide your generic design decisions.

**Comment by lemire** [comment]
> Off-topic: there's a case where it really is more important to align to cache, when atomics are involved. Atomic operations invalidate a cache line on all processors, so you want to avoid those invalidating more than one line and invalidating non-related data (so you usually try to fit more related data in the same line so it's really invalid for others anyway).

My blog post does allude to this issue. In a multithreaded context, being aware of cache lines can pay in specific instances.

**Comment by Oppen** [comment]
Oh, I thought `memcpy`ing the whole contents at once.

**Comment by Oppen** [comment]
Coming back to this. I think an option would be to:
1. Convert from portable to frozen in a buffer, without creating an intermediate bitmap;
2. Create a view with `FrozenView`;
3. Use [this trick](https://github.com/RoaringBitmap/gocroaring/commit/65d5249a532d6a8db9ad15cdee6f50896372d4ad) to keep the frozen buffer alive.

It's a bit convoluted, requires either a `[]byte` or `io.ReaderAt` (for flexibility) to be passed and makes a few single big allocations, but avoids creating a lot of little objects and recycles some code. Opinions?

**Comment by GiedriusS** [comment]
Is there still interest in this? Could I work on it?

If we are immediately performing intersection/merging after `rb.ReadFrom()`, it doesn't make sense for us to allocate gigabytes of memory potentially. Instead, we could allocate memory for a maximum of one container. That would be enough, right?

**Comment by lemire** [comment]
@GiedriusS Yes. There is still much work to do.

> Instead, we could allocate memory for a maximum of one container.

We support frozen views, see

https://github.com/RoaringBitmap/roaring/blob/a550de606142731fe523b04e108bafcf8cb5f858/serialization_littleendian.go#L298

See also discussions above.

It is deliberately not very well documented because it is not very well tested (in our project, although it is used in production by some).
